core/main/src/broker/endpoint_broker.rs

```diff
@@ -0,0 +1,598 @@
+    pub async fn send_error(&self, request: BrokerRequest, error: RippleError) {
+        let value = serde_json::to_value(JsonRpcError {
+            code: JSON_RPC_STANDARD_ERROR_INVALID_PARAMS,
+            message: format!("Error with {:?}", error),
+            data: None,
+        })
+        .unwrap();
+        let data = JsonRpcApiResponse {
+            jsonrpc: "2.0".to_owned(),
+            id: Some(request.rpc.ctx.call_id),
+            error: Some(value),
+            result: None,
+            method: None,
+            params: None,
+        };
+        let output = BrokerOutput { data };
+        if let Err(e) = self.sender.send(output).await {
+            error!("couldnt send error for {:?}", e);
+        }
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct BrokerContext {
+    pub app_id: String,
+}
+
+#[derive(Debug, Clone)]
+pub struct BrokerOutput {
+    pub data: JsonRpcApiResponse,
+}
+
+impl BrokerOutput {
+    pub fn is_result(&self) -> bool {
+        self.data.result.is_some()
+    }
+
+    pub fn get_event(&self) -> Option<u64> {
+        if let Some(e) = &self.data.method {
+            let event: Vec<&str> = e.split('.').collect();
+            if let Some(v) = event.first() {
+                if let Ok(r) = v.parse::<u64>() {
+                    return Some(r);
+                }
+            }
+        }
+        None
+    }
+}
+
+impl From<CallContext> for BrokerContext {
+    fn from(value: CallContext) -> Self {
+        Self {
+            app_id: value.app_id,
+        }
+    }
+}
+
+impl BrokerSender {
+    // Method to send the request to the underlying broker for handling.
+    pub async fn send(&self, request: BrokerRequest) -> RippleResponse {
+        if let Err(e) = self.sender.send(request).await {
+            error!("Error sending to broker {:?}", e);
+            Err(RippleError::SendFailure)
+        } else {
+            Ok(())
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct EndpointBrokerState {
+    endpoint_map: Arc<RwLock<HashMap<String, BrokerSender>>>,
+    callback: BrokerCallback,
+    request_map: Arc<RwLock<HashMap<u64, BrokerRequest>>>,
+    extension_request_map: Arc<RwLock<HashMap<u64, ExtnMessage>>>,
+    rule_engine: RuleEngine,
+}
+
+impl EndpointBrokerState {
+    pub fn get(tx: Sender<BrokerOutput>, rule_engine: RuleEngine) -> Self {
+        Self {
+            endpoint_map: Arc::new(RwLock::new(HashMap::new())),
+            callback: BrokerCallback { sender: tx },
+            request_map: Arc::new(RwLock::new(HashMap::new())),
+            extension_request_map: Arc::new(RwLock::new(HashMap::new())),
+            rule_engine,
+        }
+    }
+
+    fn get_request(&self, id: u64) -> Result<BrokerRequest, RippleError> {
+        let result = { self.request_map.read().unwrap().get(&id).cloned() };
+        if result.is_none() {
+            return Err(RippleError::InvalidInput);
+        }
+
+        let result = result.unwrap();
```

Comment:
Can be re-written to avoid unwrapping.

Reply:
To avoid concurrent lock scenarios between the read and the write(cleanup for request/response cases). result is checked for `is_some` before the unwrap so it is still safe.

Reply:
Yes but we've purposefully been removing this style from the codebase I think in part to satisfy clippy. There was a big effort to clean-up all repos.

Reply:
There are no clippy errors for this particular use case or the PR based on the GitHub Actions run. Having separate read and write blocks is necessary for the Lock safety.